stream_alert/rule_processor/rules_engine.py

```diff
@@ -365,33 +365,100 @@ def process(self, input_payload):
-
-                if types_result:
-                    record_copy = record.copy()
-                    record_copy[NORMALIZATION_KEY] = types_result
-                    if self._threat_intel and not normalized_record_appended:
-                        # A copy of payload which includes payload metadata.
-                        # The payload metadata includes log_source, type, service,
-                        # and entity. The metadata will be returned to along with
-                        # normalized record for threat detection.
-                        payload_copy = copy(input_payload)
-                        payload_copy.pre_parsed_record = record_copy
-                        payload_copy.records = None
-                        payload_copy.raw_record = None
-                        normalized_records.append(payload_copy)
-                        normalized_record_appended = True
+                    # When rule 'datatypes' option is defined, rules engine will
+                    # apply data normalization to all the record.
+                    self._apply_normalization(record, normalized_records, rule, payload, alerts)
                 else:
-                    record_copy = record
-                # rule analysis
-                self.rule_analysis(record_copy, rule, payload, alerts)
+                    self.rule_analysis(record, rule, payload, alerts)
 
         return alerts, normalized_records
 
+    def _apply_normalization(self, record, normalized_records, rule, payload, alerts):
+        """Apply data normalization to current record
+
+        Args:
+            record (dict): The parsed log w/wo data normalization.
+            normalized_records (list): Contains a list of payload objects which
+                have been normalized.
+            rule (namedtuple): Contains alerting logic.
+            payload (namedtuple): Contains parsed logs.
+            alerts (list): A list of triggered alerts.
+        """
+        normalized, normalized_payload = self._is_normalized(record,
+                                                             normalized_records,
+                                                             rule.datatypes)
+
+        record_copy = None
+        if normalized:
+            # If the record has been normalized, use normalized record copy
+            record_copy = normalized_payload.pre_parsed_record
+        else:
+            # If the record has been normalized, apply normalization logic to the record
+            # and insert normalization result to the record.
+            types_result = self.match_types(record,
+                                            payload.normalized_types,
+                                            rule.datatypes)
+
+            if types_result:
+                record_copy = record.copy()
+
+                # Insert normalization result to the record copy.
+                record_copy[NORMALIZATION_KEY] = types_result
+
+                if self._threat_intel:
```

Comment:
Is there a reason we need to couple Threat Intel to Data Normalization? This essentially means a user could not use normalization on its own

Reply:
A user can use normalization on its own by just define `datatypes` option in the Rule. Threat Intel is based on Data Normalization so threat intel will only be triggered when the normalization key existed.

Reply:
> A user can use normalization on its own by just define datatypes option in the Rule.

But the normalized record is only added to the `normalized_records` list if `self._threat_intel` is True, meaning it's tightly coupled to having Threat Intelligence enabled

Reply:
Wouldn't we want the `self._threat_intel` check to happen outside the scope of normalization to get the benefits of both?

Reply:
This block is only to add normalized records to a list. So it has to be happened right after data normalization. The list contains normalized records will be used when threat intel enabled in [handler.py](https://github.com/airbnb/streamalert/blob/master/stream_alert/rule_processor/handler.py#L125)